speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/SocketBlackBoxTests.h

```diff
@@ -0,0 +1,158 @@
+#pragma once
+#include <gtest/gtest.h>
+#include "SocketBlackBox.h"
+#include "mocks/DriveSituationMock.h"
+#include <limits>
+#include "../rppUtils/Random.hpp"
+#include <thread>
+#include "ClientSocket.h"
+#define TEST_BUFFER_SIZE 512
+
+/// @brief The black box side of the test, as these tests have to run in parallel
+void BlackBoxSide()
+{
+	SocketBlackBox<DriveSituationMock> bb;
+	CREATE_DRIVE_SITUATION_MOCK;
+	DriveSituationMock exampleSituation = GetExampleDriveSituation();
+	DriveSituationMock situations[2] {mock,exampleSituation};
+	bb.Initialize(mock,situations,2);
+	DecisionTuple decisions;
+	ASSERT_FALSE(bb.GetDecisions(mock, decisions));
+	std::this_thread::sleep_for(std::chrono::milliseconds(30));
+	ASSERT_TRUE(bb.GetDecisions(exampleSituation, decisions));
+	ASSERT_ALMOST_EQ(decisions.m_steerDecision.m_steerAmount,1.0f,0.1f);
+	ASSERT_ALMOST_EQ(decisions.m_brakeDecision.m_brakeAmount,2.0f,0.1f);
+	bb.Shutdown();
+}
+
+/// @brief					 Tests if the parsed drivesituation is the same as the target
+/// @param  p_driveSituation Parsed drive situation
+/// @param  p_target		 The target
+void TestDriveSituation(std::vector<std::string>& p_driveSituation,DriveSituationMock p_target)
+{
+	ASSERT_ALMOST_EQ(stof(p_driveSituation[0]), p_target.GetCarInfo().Speed(),0.01);
+	ASSERT_ALMOST_EQ(stof(p_driveSituation[1]), p_target.GetCarInfo().TopSpeed(),0.01);
+	ASSERT_EQ(stoi(p_driveSituation[2]), p_target.GetCarInfo().Gear());
+	ASSERT_EQ(stoi(p_driveSituation[3]), p_target.GetCarInfo().Headlights() ? 1 : 0);
+	ASSERT_ALMOST_EQ(stof(p_driveSituation[4]), p_target.GetPlayerInfo().SteerCmd(),0.01);
+	ASSERT_ALMOST_EQ(stof(p_driveSituation[5]), p_target.GetPlayerInfo().AccelCmd(),0.01);
+	ASSERT_ALMOST_EQ(stof(p_driveSituation[6]), p_target.GetPlayerInfo().BrakeCmd(),0.01);
+	ASSERT_ALMOST_EQ(stof(p_driveSituation[7]), p_target.GetPlayerInfo().ClutchCmd(),0.01);
+	ASSERT_EQ(stoi(p_driveSituation[8]), p_target.GetCarInfo().trackPosition.Offroad() ? 1 : 0);
+	ASSERT_ALMOST_EQ(stof(p_driveSituation[9]), p_target.GetCarInfo().trackPosition.ToMiddle(),0.01);
+	ASSERT_ALMOST_EQ(stof(p_driveSituation[10]), p_target.GetCarInfo().trackPosition.ToLeft(),0.01);
+	ASSERT_ALMOST_EQ(stof(p_driveSituation[11]), p_target.GetCarInfo().trackPosition.ToRight(),0.01);
+	ASSERT_ALMOST_EQ(stof(p_driveSituation[12]), p_target.GetCarInfo().trackPosition.ToStart(),0.01);
+	ASSERT_ALMOST_EQ(stof(p_driveSituation[13]), p_target.GetEnvironmentInfo().TimeOfDay(),0.01);
+	ASSERT_ALMOST_EQ(stof(p_driveSituation[14]), p_target.GetEnvironmentInfo().Clouds(),0.01);
+	ASSERT_ALMOST_EQ(stof(p_driveSituation[15]), p_target.GetEnvironmentInfo().Rain(),0.01);
+}
+
+/// @brief Tests an entire run of the framework
+TEST(SocketBlackBoxTests,SocketTest)
```

Comment:
This needs more documentation about what is happening and when. Since the test is running on multiple threads it is hard to keep track. It would be nice to document it in such a way that you can read this linearly.